controller/api/destination/endpoint_translator.go

```diff
@@ -110,58 +109,56 @@ func (et *endpointTranslator) sendFilteredUpdate(set watcher.AddressSet) {
+// consumption zone as the node. An endpoints consumption zone is set
+// by its Hints field and can be different than its actual Topology zone.
 func (et *endpointTranslator) filterAddresses() watcher.AddressSet {
-	if len(et.availableEndpoints.TopologicalPref) == 0 {
-		allAvailEndpoints := make(map[watcher.ID]watcher.Address)
-		for k, v := range et.availableEndpoints.Addresses {
-			allAvailEndpoints[k] = v
-		}
-		return watcher.AddressSet{
-			Addresses: allAvailEndpoints,
-			Labels:    et.availableEndpoints.Labels,
-		}
+	allAvailEndpoints := make(map[watcher.ID]watcher.Address)
+	for k, v := range et.availableEndpoints.Addresses {
+		allAvailEndpoints[k] = v
 	}
 
-	et.log.Debugf("Filtering through address set with preference %v", et.availableEndpoints.TopologicalPref)
-	filtered := make(map[watcher.ID]watcher.Address)
-	for _, pref := range et.availableEndpoints.TopologicalPref {
-		// '*' as a topology preference means all endpoints
-		if pref == "*" {
-			return et.availableEndpoints
-		}
-
-		srcLocality, ok := et.nodeTopologyLabels[pref]
-		if !ok {
-			continue
+	// If any address does not have a hint, then all hints are ignored and all
+	// available addresses are returned. This replicates kube-proxy behavior
+	// documented in the KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2433-topology-aware-hints/README.md#kube-proxy
+	for _, address := range et.availableEndpoints.Addresses {
+		if len(address.ForZones) == 0 {
+			return watcher.AddressSet{
+				Addresses: allAvailEndpoints,
+				Labels:    et.availableEndpoints.Labels,
+			}
 		}
+	}
 
-		for id, address := range et.availableEndpoints.Addresses {
-			addrLocality := address.TopologyLabels[pref]
-			if addrLocality == srcLocality {
+	// Each address that has a hint matching the node's zone should be added
+	// to the set of addresses that will be returned.
+	et.log.Debugf("Filtering through addresses that should be consumed by zone %s", et.nodeTopologyZone)
+	filtered := make(map[watcher.ID]watcher.Address)
+	for id, address := range et.availableEndpoints.Addresses {
+		for _, zone := range address.ForZones {
+			if zone.Name == et.nodeTopologyZone {
 				filtered[id] = address
 			}
 		}
-
-		// if we filtered at least one endpoint, it means that preference has been satisfied
-		if len(filtered) > 0 {
-			et.log.Debugf("Filtered %d from a total of %d", len(filtered), len(et.availableEndpoints.Addresses))
-			return watcher.AddressSet{
-				Addresses: filtered,
-				Labels:    et.availableEndpoints.Labels,
-			}
+	}
+	if len(filtered) > 0 {
+		et.log.Debugf("Filtered from %d to %d addresses", len(et.availableEndpoints.Addresses), len(filtered))
+		return watcher.AddressSet{
+			Addresses: filtered,
+			Labels:    et.availableEndpoints.Labels,
 		}
 	}
 
-	// if we have no filtered endpoints or the '*' preference then no topology pref is satisfied
+	// If there were no filtered addresses then the empty set is returned.
```

Comment:
I think safeguard 5 here means that we should return the full address set: https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/#safeguards

Reply:
Yep good catch.